core/src/graph/key_manager.rs

```diff
@@ -0,0 +1,283 @@
+use crate::{
+	dsnp::{
+		api_types::{DsnpKeys, KeyData, PageHash},
+		dsnp_types::{DsnpPublicKey, DsnpUserId},
+		reader_writer::{DsnpReader, DsnpWriter},
+	},
+	frequency::Frequency,
+};
+use anyhow::{Error, Result};
+use dryoc::keypair::StackKeyPair;
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
+
+trait PublicKeyProvider {
+	fn import_dsnp_keys(&mut self, keys: DsnpKeys) -> Result<()>;
+	fn export_new_keys(&self) -> Result<Vec<DsnpKeys>>;
+	fn get_all_keys(&self, dsnp_user_id: DsnpUserId) -> Vec<&DsnpPublicKey>;
+	fn get_key_by_id(&self, dsnp_user_id: DsnpUserId, key_id: u64) -> Option<&DsnpPublicKey>;
+	fn get_key_by_public_key(
+		&self,
+		dsnp_user_id: DsnpUserId,
+		public_key: Vec<u8>,
+	) -> Option<&DsnpPublicKey>;
+	fn get_active_key(&self, dsnp_user_id: DsnpUserId) -> Option<&DsnpPublicKey>;
+	fn add_new_key(&mut self, dsnp_user_id: DsnpUserId, public_key: Vec<u8>) -> Result<u64>;
+}
+
+trait UserKeyProvider {
+	fn import_key_pairs(&mut self, pairs: Vec<StackKeyPair>);
+	fn get_key(&self, key_id: u64) -> Option<(DsnpPublicKey, StackKeyPair)>;
+}
+
+pub struct PublicKeyManager {
+	/// keys are stored sorted by index
+	dsnp_user_to_keys: HashMap<DsnpUserId, (Vec<DsnpPublicKey>, PageHash)>,
+	new_keys: HashMap<DsnpUserId, DsnpPublicKey>,
+}
+
+pub struct UserKeyManager {
+	public_key_manager: Rc<RefCell<PublicKeyManager>>,
+	dsnp_user_id: DsnpUserId,
+	keys: Vec<StackKeyPair>,
+}
+
+impl PublicKeyProvider for PublicKeyManager {
+	/// importing dsnp keys as they are retrieved from blockchain
+	/// sorting indices since id's might not be unique but indices definitely should be
```

Comment:
Hmm... I'm wondering if it might be worth building an item key into Itemized storage, instead of just relying on positional index...
* Wouldn't really be any more computationally expensive, since we're already reading the items into a BTreeMap indexed by item (would just change index to key)
* Wouldn't involve any different storage reads/writes
* Would be able to eliminate duplicate-keyed items on-chain
* Could support "upsert" for items

Drawbacks:
* Would not be able to "update" an item in AppendOnly by re-publishing